martha/martha_v3.js

```diff
@@ -347,12 +350,45 @@ async function retrieveFromServers(params) {
     let fileName;
     let localizationPath;
     let accessUrl;
+    let passports;
 
     // `fetch` below might time out in a way that we want to report as an error. Since the timeout
     // interrupts us while we're waiting for a response, we need to capture what we were about to
     // try doing just before we do it so that we can provide that detail in the error report.
     let hypotheticalErrorMessage;
 
+    const getAccessUrl = async (args) => {
```

Comment:
I think this function is doing a little too much by auto-handling the fallback case. Consequently:

* the caller has wasted time fetching a Fence token even if it didn't need to
* errors from calling Bond and Fence will cause the Martha invocation to fail even if the passport would have worked

I'd rather this function return nothing if fetching with the passport fails and move the fallback logic to the caller.

(I first noticed this from the tests; I was surprised to see that they expected Bond to be called.)

Reply:
I don't think this should be the caller's responsibility. This is one of the purpose's of Martha: figure out the authentication.

I agree that the caller's time is wasted fetching the fence token, I will work on that.

Reply:
Unless you mean caller as in caller of `getAccessUrl` and not caller of Martha.

Reply:
Yes, sorry, I meant the caller in the context of this function, `getAccessUrl`, not the caller of Martha.